common/os_calls.c

```diff
@@ -2876,6 +2876,45 @@ g_waitpid(int pid)
 #endif
 }
 
+/*****************************************************************************/
+/* does not work in win32
+   returns exit code of child process with pid */
+int
+g_waitpid_status(int pid)
+{
+#if defined(_WIN32)
+    return -1;
+#else
+    int rv = 0;
+    int status;
+
+    if (pid <= 0)
+    {
+        rv = -1;
+    }
+    else
+    {
+        LOG(LOG_LEVEL_DEBUG, "waiting for pid %d to exit", pid);
+        rv = waitpid(pid, &status, 0);
+
+        if (rv != -1)
+        {
+            if (WIFEXITED(status))
+            {
+                rv = WEXITSTATUS(status);
+            }
+        }
+        else
+        {
+            LOG(LOG_LEVEL_WARNING, "wait for pid %d returned unknown result", pid);
+            rv = -1;
+        }
+    }
+
+    return rv;
+#endif
+}
+
```

Comment:
This function doesn't cope with a process exiting after catching a signal. This is relatively common for chansrv. In such cases it looks like rv is set to zero which is misleading.

There's no great answer here I can see without changing the interface, but a better solution might be to include (after the WIFEXITED() test):-

```c
else if (WIFSIGNALED(status))
{
    /* Specify a value not representable in 8 bits */
    rv = -128 - WTERMSIG(status);
}
```

The logic is that `WEXITSTATUS()`returns an 8 bit value (either 0-255 or -128-127), so we can at least see a signal for what it is. This can probably be improved on.

Reply:
Since `g_waitpid_status()` is a new function I just changed the interface to return a struct with the return code and signal code.